core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2026,13 +2026,15 @@ public void startRecovery(RecoveryState recoveryState, PeerRecoveryTargetService
                 break;
             case LOCAL_SHARDS:
                 final IndexMetaData indexMetaData = indexSettings().getIndexMetaData();
-                final Index mergeSourceIndex = indexMetaData.getMergeSourceIndex();
+                final Index mergeSourceIndex = indexMetaData.getResizeSourceIndex();
                 final List<IndexShard> startedShards = new ArrayList<>();
                 final IndexService sourceIndexService = indicesService.indexService(mergeSourceIndex);
-                final int numShards = sourceIndexService != null ? sourceIndexService.getIndexSettings().getNumberOfShards() : -1;
+                final Set<ShardId> requiredShards = IndexMetaData.selectRecoverFromShards(shardId().id(),
+                    sourceIndexService.getMetaData(), indexMetaData.getNumberOfShards());
```

Comment:
sourceIndexService could be null? (you're doing the null check in the next line)